### PR TITLE
Rewrite join! and try_join! as procedural macros

### LINKED ISSUE #1782: Rewrite join! and try_join! macros as procedural macros
https://github.com/rust-lang/async-book/pull/28#discussion_r311183878

Related: #1363

### PR DESCRIPTION
**Pull request by taiki-e**
Closes #1782

r? @cramertj

### CODE REVIEW

File: futures-util/Cargo.toml [hunk 1]
```diff
@@ -27,13 +27,15 @@ cfg-target-has-atomic = ["futures-core-preview/cfg-target-has-atomic"]
 sink = ["futures-sink-preview"]
 io = ["std", "futures-io-preview", "memchr"]
 channel = ["std", "futures-channel-preview"]
+join-macro = ["async-await", "futures-join-macro-preview", "proc-macro-hack", "proc-macro-nested"]
```
Comment (taiki-e) [inline futures-util/Cargo.toml]: For almost the same reason as `select!`, this is optional in `futures_util` (see https://github.com/rust-lang-nursery/futures-rs/issues/1701).

**Comment by cramertj** [comment]
Wow, this is awesome! I have so much appreciation for your eagerness and quick work on all of this. Thanks, @taiki-e!

File: futures-util/src/async_await/join_mod.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+macro_rules! document_join_macro {
+    ($join:item $try_join:item) => {
+        /// Polls multiple futures simultaneously, returning a tuple
+        /// of all results once complete.
+        ///
+        /// While `join!(a, b)` is similar to `(a.await, b.await)`,
+        /// `join!` polls both futures concurrently and therefore is more efficent.
+        ///
+        /// This macro is only usable inside of async functions, closures, and blocks.
+        /// It is also gated behind the `async-await` feature of this library, which is
```
Comment (Nemo157) [inline futures-util/src/async_await/join_mod.rs]:
```suggestion
        /// It is also gated behind the `join-macro` feature of this library, which is
```

Comment (Nemo157) [reply futures-util/src/async_await/join_mod.rs]: Oh, it depends on what "this library" is referring to...

Comment (taiki-e) [reply futures-util/src/async_await/join_mod.rs]: Yeah. So, I would like to remove these descriptions and change them to use `doc(cfg(feature = ".."))` instead in the near future.

**Pull request merged**